### PR TITLE
Add reverse list

### PR DESCRIPTION
**Pull request by TheLimeGlass**
**Target Minecraft versions:** any

**Description:**
Adds the ability to reverse a list.

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			Collections.reverse(reversed);
+		} catch (IllegalArgumentException ex) { // In case elements are not comparable
+			Skript.error("Tried to reverse a list, but some objects are not comparable!");
+		}
+		return reversed.toArray();
+	}
+	
+	@Override
+	public Class<? extends Object> getReturnType() {
+		return Object.class;
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: Any reason why it couldn't be `list.getReturnType()` like here? https://github.com/SkriptLang/Skript/blob/8c26dca2f0520bbac0e1817c8af58229b1a05669/src/main/java/ch/njol/skript/expressions/ExprRandom.java#L79
Only wondering

File: src/main/java/ch/njol/skript/expressions/ExprShuffledList.java [hunk 2]
```diff
@@ -62,14 +62,14 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 	
 	@Override
 	@Nullable
-	protected Object[] get(Event e) {
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprShuffledList.java]: This should be `e`, @bensku even specifically changed it from `event` to `e` in one of my commits

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	
+	@Override
+	public boolean isSingle() {
+		return false;
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return "reversed list";
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: in `ExprRandom` it calls toString on the list too, so `return "reversed list " + list.toString(e, debug)` would probably also be better here

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	
+	@Override
+	@Nullable
+	protected Object[] get(Event event) {
+		Object[] origin = list.getArray(event);
+		List<Object> reversed = Arrays.asList(origin.clone());
+		try {
+			Collections.reverse(reversed);
+		} catch (IllegalArgumentException ex) { // In case elements are not comparable
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: What does comparing have to do with reversing a list?

File: src/main/java/ch/njol/skript/expressions/ExprShuffledList.java [hunk 5]
```diff
@@ -62,14 +62,14 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 	
 	@Override
 	@Nullable
-	protected Object[] get(Event e) {
-		Object[] origin = list.getAll(e);
+	protected Object[] get(Event event) {
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprShuffledList.java]: Same here, should be `e`

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java (hunk 4 above)
Comment (xXAndrew28Xx) [reply src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: ^ Collections#reverse does not throw an IllegalArgumentException if the elements are not comparable as it does not compare any elements. Reversing something doesn't require comparing the elements unlike sorting.

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java (hunk 3 above)
Comment (Blueyescat) [reply src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: "reversed " + list.toString(e, debug)

File: src/main/java/ch/njol/skript/expressions/ExprReversedList.java (hunk 4 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/expressions/ExprReversedList.java]: maybe
```java
Collections.reverse(((Map) list.getRaw(e)).values());
```
could be faster

**Comment by Blueyescat** [comment]
No no no string representations shouldn't be like that (`reversed list: <list>`), it should be just `reversed <list>`.. They will be used in error/warning messages and what you used is not user friendly.

File: src/main/java/ch/njol/skript/expressions/ExprRandom.java [hunk 6]
```diff
@@ -47,13 +47,24 @@
...
 public class ExprRandom extends SimpleExpression<Object> {
+
 	static {
 		Skript.registerExpression(ExprRandom.class, Object.class, ExpressionType.COMBINED, "[a] random %*classinfo% [out] of %objects%");
 	}
 	
 	@SuppressWarnings("null")
 	private Expression<?> expr;
 	
+	@Override
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprRandom.java]: Why did you move this above, it's always `init` -> `get` -> `isSingle` and `getReturnType` -> `toString`, unless there's other stuff

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java [hunk 7]
```diff
@@ -48,13 +49,32 @@
...
+		if (exprs[0] instanceof Variable<?>) {
+			Variable<?> variable = ((Variable<?>) exprs[0]);
+			if (!variable.isList()) {
+				Skript.error("You may only sort list variables, not single variables.");
+				return false;
+			}
+			list = exprs[0].getConvertedExpression(Object.class);
+			return true;
+		}
+		if (!Comparable.class.isAssignableFrom(type)) {
```
Comment (Nicofisi) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: Did you check how your fix works with the code example from #1433?

**Comment by Nicofisi** [comment]
Also, I know that strictly speaking Skript doesn't have any code standards yet set, but I think it would be nice if you added empty lines *after* the last bracket (GitHub ever warns about the lack of them), and removed the ones before the last bracket 
image

File: src/main/java/ch/njol/skript/expressions/ExprRandom.java (hunk 6 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprRandom.java]: No it's not. It is always the way I put it.

It's the order asending from the simple expression class.

**Comment by TheLimeGlass** [comment]
I did add empty line breaks to some. That's just untouched.

File: src/main/java/ch/njol/skript/expressions/ExprRandom.java (hunk 6 above)
Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/expressions/ExprRandom.java]: But that isn't the convention so can you please move them where they belong?

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java (hunk 7 above)
Comment (Snow-Pyon) [reply src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: I meant, variables. It isn't supposed to work only with variables.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: What do you mean? There is variable support.

**Comment by Snow-Pyon** [comment]
Are you going to fix the requested changes?

**Comment by TheLimeGlass** [comment]
I did

File: src/main/java/ch/njol/skript/expressions/ExprSortedList.java (hunk 7 above)
Comment (xXAndrew28Xx) [inline src/main/java/ch/njol/skript/expressions/ExprSortedList.java]: Reverse list **does not** need sorting (Comparable) to be implemented in the elements.

**Comment by Snow-Pyon** [comment]
Tried to do the changes myself before merging but it looks like something got fucked up, going to merge it manually.

Sorry for the inconveniences!

**Comment by bensku** [comment]
Honestly, I'm not sure. Something with Git went *really* wrong.
